Modify aliyun_ftasr sdk

Lizerui9926 opened the pull request:
Modify the method of calling aliyun_ftasr sdk to parallel to improve efficiency.

dophist commented on models/aliyun_ftasr_api_zh/asr_sdk.py:
```diff
@@ -122,6 +124,22 @@ def fileTrans(akId, akSecret, appKey, fileLink) :
+        text = ''
+        text = fileTrans(access_key_id, access_key_secret, APPKEY, audio)
+
+        trans.write(key + '\t' + text + '\n')
```
* trans is out-of-scope of asr() ? I don't see asr() as a closure to capture that.
* do we need locks to avoid concurrency contention ? 

Please take your PR with care, always validate your submission.  It's impossible for me to look into the code of all model-images inside model-zoo, and each failing benchmarking run *WILL* cost time & money.

Lizerui9926 commented:
Thank you for looking into this PR.
1.Sorry, I don't quite understand your first question. We encapsulate _trans.write_ into a function to achieve parallel calls. We modified our code and changed the single-process sequential execution to multi-process parallel execution in order to improve recognition speed.
2.Our recording file recognition is an offline recognition service, and there is no limit to the number of concurrent. And We have limited the number of multi-processes to half the number of cpu cores in the new commit.

We have performed multiple validations of our submission on the UNLOCKED test sets to ensure proper execution.

dophist commented:
I just tried to test the already-merged aliyun_ftasr_api_zh, got:

log.SBI:
```
Invalid line: e5oipIfM49I__20190201_CCTV_1  Error: accessKeyID and ecsUrl are both empty
Invalid line: e5oipIfM49I__20190201_CCTV_104    Error: accessKeyID and ecsUrl are both empty
```

your submission is not working. Please debug, fix, and *test*.

For this particular PR (parallel improvement), my questions are:
1. the `trans` variable in asr() seems to be undefined to me, it is neither in global scope, nor in local scope of asr(), and I don't see there is a lambda capturing it.  I might be wrong, but please explain. 
2. you are writing a single file in a multiprocess program, even you flush() right away, there might still be IO contentions.  Without locks or queues, the result can be messed up. If not, that's just lucky to me. Each benchmark costs us ~300 Yuan, we can't rely on luck.
3. During the benchmarking, there are almost 10 models running in parallel, Aliyun is one of them.  It's problematic for one model to use half of those CPUs.  Please make the number of CPUs configurable and with default value of 5.

dophist commented:
And one more thing, does this PR mean that your submission will fork thousands of subprocesses during the benchmarking, given a test set with thousands of utterances?

I would not do this in practice, you may consider splitting the test set into 10 parts or so, and run each part in a separate process with its own IO traffics, and merge them in the end.

Lizerui9926 commented:
This problem of "Error: accessKeyID and ecsUrl are both empty" may be that the oss configuration file is not generated in the home directory. (The default path of the configuration file is: /home/user/.ossutilconfig).  We will first roll back to the code that has been merged, and  fix and debug based on it.

Lizerui9926 commented:
We modified the pipeline to specify the configuration file path, and added the -c option to specify the configuration file every time when the **_ossutil64_** command is used. We verified on our machine, generated the oss configuration file in the specified path, and successfully got the return result.

Lizerui9926 commented:
Thanks for your suggestion. Now we split the test set into several parts for testing, write the asr results into txt file respectively, and finally merge them into _raw_rec.txt_.

dophist commented on models/aliyun_ftasr_api_zh/SBI:
```diff
@@ -8,11 +8,29 @@ fi
-./asr_sdk.py $oss_scp $dir/raw_rec.txt
+
+# split test set 
+awk -vf=${part_nums} -vl="`wc -l $oss_scp`" 'BEGIN{p=int(l/f);q=(l%f);for(n=1;n<=f;n++)a[n]=n*p+((n<=q)?++x:x)}{if(NR>a[i])i++;print > "'$oss_scp'."i}' $oss_scp
```
Here you append $oss_scp  with split indexes as temporary paths, according to spec, you are not allowed to write temp files outside "result dir" (i.e. $dir). please put them explicitly inside $dir

dophist commented on models/aliyun_ftasr_api_zh/SBI:
```diff
@@ -8,11 +8,29 @@ fi
+part_nums=10
 
 # download ossutil64 tools
 wget -O ossutil64 https://gosspublic.alicdn.com/ossutil/1.7.13/ossutil64
```
I just realized that there might be multiple model-images running in parallel during the benchmark, downloading is not isolated, it may have side effect to other benchmarking process. Could you please:
1. use a pre-downloaded ossutil binary in model-image (as your initial submission)
2. and add a MD5 check in SBI to ensure the binary is consistent with expected MD5 signature
3. and add a comment in the code as well, telling where to find this MD5 signature(e.g. Official Aliyun webpage)

dophist commented on models/aliyun_ftasr_api_zh/SBI:
```diff
@@ -8,11 +8,29 @@ fi
+cat $dir/sdk_log.* > $dir/sdk_log
+rm -rf $dir/sdk_log.*
+rm -rf $dir/raw_rec.txt.*
+rm -rf ${oss_scp}.*
```
same as above

Lizerui9926 commented:
Now, we specify the _$oss_scp_(_${dir}/wav_oss.scp_) and other temp files in the _$dir_ and write the _myossconfig_ file to $dir as well. We add _ossutil64_ binary and its MD5 file. If MD5 check does not meet expectations, the pipeline will be exited.

dophist commented:
Is there a way to avoid the generation of myossconfig? Like, credentials (AK and AK secret) are passed through ossutil's arguments directly, that would be better.  Because I don't expect the result dir containing any "secret" stuff, the contents of result dir should be shareable to public, someday. Thanks

Lizerui9926 commented:
The default path of _myossconfig_ is /home/user/.ossutilconfig, but if the path is not specified, the default config file may not be generated, resulting in "Error: accessKeyID and ecsUrl are both empty" problem. Maybe we can delete _myossconfig_ after executing the pipeline every time, is this feasible? Thanks.

dophist commented:
Does `ossutil` have options for AK and AK secret?

dophist commented:
One more thing, I noticed oss://speechiotest/ bucket is used across all tests.  Can we improve to use a unique bucket name so that different benchmarks will never have storage overlaps, may be something like oss://speechiotest_`unique_id`, where the unique_id is generated from test set id, or even a uuid-like identifier seeded by system time.

Lizerui9926 commented:
Thanks for you suggestion. We add system time as unique_id and modify the bucket name. And we remove _ossutil64 config_ command and add the -i and -k option for AK and AK secret. We verify it on our machine and it is successful.

dophist commented:
Thanks. Seems good to me now. I'll merge this PR right before this season's benchmark (probably at Aug 20th), feel free to improve or fix this PR before that.
